### ISSUE TITLE
Thunderbird: The filter lets slip through some things that are blocked on Firefox

### ISSUE THREAD

**Issue by M-Reimer** [gorhill/uBlock#897]
Unfortunately I'm not able to debug this on my own.

I have the uBlock version with initial Thunderbird support in daily use to read several RSS feeds. On this feed: http://feeds2.feedburner.com/stadt-bremerhaven/dqXM I recognized, that sometimes advertisements from "doubleclick.net" are still visible. I used "DOM Inspector" to find out the actual URL and copied it to Firefox where it is blocked without any problems.

I use the exactly same filter settings in Firefox and Thunderbird.

If I compare the result of the Logger, built into uBlock, between Firefox and Thunderbird, then I see that the "doubleclick.net" items are missing in Firefox. Seems like something in earlier stage is blocked and so the doubleclick.net content isn't even requested.

In Thunderbird, I have doubleclick.net entries in the Logger, but they are listed as "cosmetic filtering".

So something is going wrong here. Currently it seems like it is a bad idea to use uBlock Origin in Thunderbird, if someone wants to use it for security reasons, as some things, that are on the filter lists, aren't really filtered out.

I don't know why this happens. I expected that the backend (content policy to prevent loading) works in Thunderbird just as well as in Firefox, but this seems to be wrong. Would be great to get some help with this one...

**Comment by gorhill** [comment]
I see a lot of network requests being done from within the behind-the-scene context, i.e. uBO can't associate them to a specific context. You can make behind-the-scene network requests become filtered by:

- Remove the whitelisting of `behind-the-scene`.
- Enable "I am an advanced user" mode.

This is not ideal though, as some filters are designed to work from within specific context only.

**Comment by RoxKilly** [comment]
Suppose I have the filter `* doubleclick.net * block` in `My Rules` to express that I never want to connect to doubleclick.  If a `behind-the-scene` request is made to that domain, will it be blocked?  In that case, the context shouldn't matter given the rule above, but you have explained [here](https://github.com/gorhill/uBlock/wiki/Overview-of-uBlock's-network-filtering-engine) and [here](https://github.com/gorhill/uBlock/wiki/Dynamic-filtering:-precedence) that whitelisting has precedence over filtering. Is this true even when the context in the rule is the wildcard character `*`?

**Comment by gorhill** [comment]
> Is this true even when the context in the rule is the wildcard character `*`?

Yes. Whitelisting means "**disabling** uBlock's filtering for the current site" (or whatever matches the whitelist directive).

**Comment by M-Reimer** [comment]
@gorhill how exactly does uBlock Origin detect those contexts? Or: What is the minimum, that I have to implement, to fix this problem?

**Comment by gorhill** [comment]
> how exactly does uBlock Origin detect those contexts?

uBO keeps an internal map of tab id/frame id to URL. Each network request is tagged with the tab id/frame id from where it comes, so uBO looks up the proper URL according to the tab id/frame id. That URL is the context. When no tab id can be looked up, the network request is categorized as behind-the-scene.

**Comment by bhabba** [comment]
I have removed ``behind-the-scene`` from the whitelist and also removed the pre-configured rules ``behind-the-scene * 3p noop`` and ``behind-the-scene * 3p-frame noop`` (although this shouldn't have any effect unless I would have set other relevant dynamic rules if I understand it right). There are no dynamic rules left. I have subscribed to the default filter lists plus some more.

Now as you can see in the logger screenshot below some requests get successfully blocked (gg.google.com) while others get through (doubleclick.net). This shouldn't be, right?

![thunderbird_logger](https://cloud.githubusercontent.com/assets/3280609/12868664/465359de-cd0b-11e5-8807-1db96fba5c8c.png)

uBO 1.5.5 with Thunderbird 38.5.1 on Ubuntu 14.04.3 x64

**Comment by gorhill** [comment]
> This shouldn't be, right?

Why?

**Comment by bhabba** [comment]
> Why?

Should I conclude from your counterquestion that this is the expected behaviour with my settings? I don't understand why doubleclick.net isn't blocked.

**Comment by gorhill** [comment]
> I don't understand why doubleclick.net isn't blocked.

But it shows right there in the logger why, there is an exception filter -- in green. Click that filter to find out where it comes from.

**Comment by bhabba** [comment]
Thank you, you're absolutely right! I just didn't want to believe that there really could be an exception rule for "doubleclick.net/instream/ad_status.js".

The exception rule came from "Fanboy's Enhanced Tracking List" as shown below:

![doubleclick filter](https://cloud.githubusercontent.com/assets/3280609/12869393/988dc0d2-cd1e-11e5-8885-ec1198e8b755.png)
The .js file loaded there consists of only one line: ``window.google_ad_status = 1;``
I have no clue what this does.
To block everything from doubleclick.net I would have to add the static filter ``||doubleclick.net$important`` I guess.

Anyway, so to get uBO properly working with Thunderbird the only thing to do is removing ``behind-the-scene`` from the whitelist? Couldn't this be done automatically during installation (of course only if thunderbird is detected as installation target)?

**Comment by gorhill** [comment]
If the content is so simple as this, I have no problem creating a neutered script for this one.

**Comment by gorhill** [comment]
What would help is if you can [inquire why this filter was added](https://forums.lanik.us/viewforum.php?f=62), maybe it's obsolete by now?

**Comment by bhabba** [comment]
> What would help is if you can inquire why this filter was added, maybe it's obsolete by now?

According to fanboy ["it's an adblock tracking script"](https://forums.lanik.us/viewtopic.php?f=62&t=28427&sid=765bb1230dea7e7489ae2c79b87b6a26#p86455). Whatever that exactly means... :no_mouth:

**Comment by IDKwhattoputhere** [comment]
Why didn't you ask him what it means? It's most likely a script with which Google keeps track of which users use an adblocker.

**Comment by gorhill** [comment]
Exception filters are used to un-break a feature on a site, I just wonder what feature was broken by not having this filter. This will allow me to test the neutered script.

**Comment by IDKwhattoputhere** [comment]
https://forums.lanik.us/viewtopic.php?p=86457#p86457
Waiting for an answer.

**Comment by bhabba** [comment]
answer from [fanboy](https://forums.lanik.us/viewtopic.php?f=62&t=28427&sid=765bb1230dea7e7489ae2c79b87b6a26#p86472):

> If blocked, doubleclick will know adblock is being used. And there are more sites doing this:
> 
> https://hg.fanboy.co.nz/file/c247148923b3/enhancedstats-addon.txt#l663
> 
> Basically countering adblock measuring/tracking.

**Comment by ryanbr** [comment]
The filter won't cause any ads or popups. its just monitoring adblock users.

**Comment by IDKwhattoputhere** [comment]
@ryanbr If I read @gorhill's solution correctly it will let Google detect that there's no adblocker installed without the need to connect to a DoubleClick server.

**Comment by ryanbr** [comment]
Connecting to the doubleclick server isn't evil (in this case), its a script intending to detect adblock users blocking it

**Comment by gorhill** [comment]
> Basically countering adblock measuring/tracking

Ok I see. Then at this point a pro-argument of having the neutered script would be to have that countering being enabled by default (currently it's an experimental filter though) -- a good thing for those not subscribing to Enhanced Tracking List.

**Comment by bhabba** [comment]
@gorhill 
The redirection 

``||doubleclick.net/instream/ad_status.js$script,redirect=doubleclick.net/instream/ad_status.js,important``

you're using for this let's the doubleclick server connect to itself instead of us, right?

Why not add the same for the [other domains in Fanboy's Enhanced Tracking List](https://hg.fanboy.co.nz/file/c247148923b3/enhancedstats-addon.txt#l668) that do adblock measuring/tracking?:

*    668 @@/plugins/ad-blocking-detector/*$script,~third-party
*    669 @@||9c9media.com^*/advertisement.js
*    670 @@||buzzfed.com^*/ads.js
*    671 @@||canada.com^*/advertisement.js
*    672 @@||cdnslate.com/wp-srv/ad/*/loader.min.js
*    673 @@||classistatic.com^*/advertising.js
*    674 @@||comeadvertisewithus.com^*/ads.js
*    675 @@||dailymail.co.uk/adstest2/$script
*    677 @@||espncdn.com^*/ads.js
*    678 @@||exactly0r.com/advertising.js
*    679 @@||forums.qrz.com/js/advertisement.js
*    680 @@||fsdn.com/con/js/adframe.js
*    681 @@||ft-static.com^*/ads.js$domain=ft.com
*    682 @@||gamespot.com/js/ads.js
*    683 @@||motherjones.net/adtracking.js
*    684 @@||motherless.com/scripts/site_ads.js
*    685 @@||oddee.com/js/advertising.js
*    686 @@||ookla.com/www/test.js
*    687 @@||ora.tv^*/adframe.js
*    688 @@||photobucket.com^*/ads.js
*    689 @@||pornrabbit.com/static/advertisement.js
*    690 @@||ps4news.com/ad/adframe.js
*    691 @@||smartadcheck.de/js/*/ad/$script
*    692 @@||theatlantic.com^*/advertisement.js
*    693 @@||tnwcdn.com/assets/js/advertising.js$domain=thenextweb.com
*    694 @@||upworthy.com/assets/ads/$script
*    695 @@||vice.com/assets/js/vmp/adblocker/
*    696 @@||vox-cdn.com/javascripts/lib/advertisement.

**Comment by ryanbr** [comment]
personally signaling out one domain (doubleclick) vs everything else seems a bit pointless. Then again its also pointless to counter something that is designed to stop adblock tracking.

**Comment by bhabba** [comment]
> Then again its also pointless to counter something that is designed to stop adblock tracking.

Why? It seems a reasonable goal to foil adblock tracking (or at least synthetically reduce the adblock statistics of the trackers), no?

**Comment by ryanbr** [comment]
That was the whole point of the tracking, and to fool adblock scripts to thinking you're using no Adblock software is another form of tracking. Seems strange to attempt to counter this one aspect of the list.

**Comment by ryanbr** [comment]
tl;dr If you want to be tracked, Enhanced Tracking list isn't for you.

**Comment by gorhill** [comment]
uBO in [medium mode](https://github.com/gorhill/uBlock/wiki/Blocking-mode:-medium-mode) would cause `static.doubleclick.net/instream/ad_status.js` to be blocked regardless of the exception filter in Enhanced Tracking. The redirection filter will help in such case. Sure it's one case out of many others, but given how `doubleclick.net` is ubiquitous, I believe it's worth to have this specific redirect filter given how trivial it is (one line of js code).

**Comment by bhabba** [comment]
> to fool adblock scripts to thinking you're using no Adblock software is another form of tracking

After reading the [introduction to NoScript's surrogates](https://hackademix.net/2009/01/25/surrogate-scripts-vs-google-analytics/) from it's developer Giorgio Maone I think I understand uBO's redirection principle better now.

If I get it right, the redirection gorhill uses makes the resource that calls/references ``static.doubleclick.net/instream/ad_status.js`` think you're not using an adblocker **without** leaking privacy (i.e. it doesn't let doubleclick track you).

So to me it would seem desirable to implement this behaviour for the other relevant sites in Fanboy's Enhanced Tracking List as well.

**Comment by IDKwhattoputhere** [comment]
It seems like YouTube uses a second request to test for adblockers. `https://googleads.g.doubleclick.net/pagead/id` shows up on every YT page and only responds with HTTP status code 204, nothing else. Seems like another adblock test to me since EasyList blocks doubleclick.net on YT.

**Comment by okiehsch** [comment]
@gorhill can the filter 
````
||doubleclick.net/instream/ad_status.js$script,redirect=doubleclick.net/instream/ad_status.js,important
````
be added to uBO-filters and removed from uBO-experimental?
It fixes for example https://github.com/jspenguin2017/uBlockProtector/issues/581

**Comment by gorhill** [comment]
Duplicate of #117.
